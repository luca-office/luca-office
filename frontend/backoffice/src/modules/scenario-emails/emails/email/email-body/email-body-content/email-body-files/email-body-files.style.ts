import {css} from "@emotion/react"
import {inputHeight, spacingSmall, textEllipsis} from "shared/styles"

export const emailStyle = {
  iconTitleColumn: css({
    flexBasis: "90%"
  }),
  deleteColumn: css({
    display: "flex",
    justifyContent: "flex-end",
    paddingRight: 0,
    minWidth: inputHeight
  }),
  addButton: css({
    width: inputHeight,
    height: inputHeight
  }),
  addButtonIcon: css({
    margin: 0
  }),
  nameColumnContent: css(textEllipsis, {
    display: "grid",
    gridTemplateColumns: "auto 1fr",
    gridGap: spacingSmall,
    alignItems: "center"
  }),
  deleteButton: css({
    paddingRight: spacingSmall
  }),
  deleteButtonInner: css({
    background: "transparent"
  })
}
