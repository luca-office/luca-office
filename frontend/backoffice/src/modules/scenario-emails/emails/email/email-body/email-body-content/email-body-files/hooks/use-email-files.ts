import React from "react"
import {FileUsageType, Relevance, TextDocumentCreation} from "shared/graphql/generated/globalTypes"
import {useCreateFile, useDeleteFileFromScenario, useFilesForScenario} from "shared/graphql/hooks"
import {useCreateTextDocument} from "shared/graphql/hooks/mutations/text-document/use-create-textdocument"
import {DeleteEntityHook, Email, File as AppFile, UploadBinary} from "shared/models"

export interface UseEmailFilesHook {
  readonly files: AppFile[]
  readonly filesLoading: boolean
  readonly uploadVisible: boolean
  readonly deleteFileHook: () => DeleteEntityHook
  readonly uploadBinaries: (binaries: UploadBinary[]) => void
  readonly showUpload: () => void
  readonly hideUpload: () => void
  readonly createTextDocumentFile: (textDocument: TextDocumentCreation, title: string) => void
}

export const useEmailFiles = (email: Email): UseEmailFilesHook => {
  const {id, scenarioId} = email
  const [showUpload, setShowUpload] = React.useState(false)
  const {createFile} = useCreateFile({scenarioId})
  const {files: scenarioFiles, filesLoading} = useFilesForScenario(scenarioId)
  const deleteFileFromScenario = useDeleteFileFromScenario(scenarioId)
  const files = scenarioFiles
    .map(file => file.filter(f => f.usageType === FileUsageType.Email && f.emailId === email.id))
    .getOrElse([])
  const {createTextDocument} = useCreateTextDocument({scenarioId})

  const uploadBinaries = (binaries: UploadBinary[]) => {
    Promise.all(
      binaries.map(binary =>
        createFile({
          name: binary.data.filename,
          relevance: Relevance.PotentiallyHelpful,
          tags: [],
          emailId: id,
          usageType: FileUsageType.Email,
          ...{[binary.isSpreadsheet ? "spreadsheetId" : "binaryFileId"]: binary.data.id}
        })
      )
    )
      .catch(err => {
        console.error(err)
      })
      .finally(() => setShowUpload(false))
  }

  const createTextDocumentFile = (textDocument: TextDocumentCreation, title: string) => {
    createTextDocument(textDocument).then(textDocumentOption =>
      textDocumentOption.forEach(textDocument =>
        createFile({
          name: title,
          relevance: Relevance.PotentiallyHelpful,
          tags: [],
          usageType: FileUsageType.Email,
          textDocumentId: textDocument.id,
          emailId: email.id
        })
      )
    )
  }

  return {
    hideUpload: () => setShowUpload(false),
    showUpload: () => setShowUpload(true),
    uploadVisible: showUpload,
    deleteFileHook: () => deleteFileFromScenario,
    files,
    filesLoading,
    uploadBinaries,
    createTextDocumentFile
  }
}
