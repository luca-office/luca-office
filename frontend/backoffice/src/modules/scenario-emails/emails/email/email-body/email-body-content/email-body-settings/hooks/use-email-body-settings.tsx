import * as React from "react"
import {useDispatch} from "react-redux"
import {SelectOptionCustomized} from "shared/components"
import {NotificationSeverity} from "shared/enums"
import {EmailDirectory, EmailUpdate} from "shared/graphql/generated/globalTypes"
import {AppNotification, Email} from "shared/models"
import {useLucaTranslation} from "shared/translations"
import {Option} from "shared/utils"
import {EmailUnreadIndicator} from "../../../../../../../../components"
import {EmailStatus} from "../../../../../../../../enums"
import {navigateToRouteAction} from "../../../../../../../../redux/actions/navigation-action"
import {updateNotification} from "../../../../../../../../redux/actions/ui/common-ui-action"
import {Route} from "../../../../../../../../routes"
import {getDirectoryOptions} from "../../../../../../utils/directory"

export interface UseEmailBodySettingsHook {
  readonly directoryOptions: SelectOptionCustomized[]
  readonly hideDelayOverlay: () => void
  readonly isDelayOverlayVisible: boolean
  readonly markerOptions: SelectOptionCustomized[]
  readonly moveEmail: (directory: EmailDirectory) => void
  readonly showDelayOverlay: () => void
}

export const useEmailBodySettings = (
  email: Email,
  updateEmail: (update: Partial<EmailUpdate>, onSuccess?: () => void) => Promise<Option<Email>>
): UseEmailBodySettingsHook => {
  const {t} = useLucaTranslation()
  const dispatch = useDispatch()
  const [isDelayOverlayVisible, setDelayOverlayVisible] = React.useState<boolean>(false)
  const switchDirectory = (directory: EmailDirectory) =>
    dispatch(navigateToRouteAction(Route.ScenarioEmails, {scenarioId: email.scenarioId, directory, emailId: email.id}))
  const showMoveWarning = () =>
    dispatch(
      updateNotification(
        Option.of<AppNotification>({
          severity: NotificationSeverity.Warning,
          messageKey: "notification__message_email_folder_delay"
        })
      )
    )
  const moveEmail = (directory: EmailDirectory) => {
    const successFn = () => switchDirectory(directory)

    if (
      email.directory === EmailDirectory.Inbox &&
      email.receptionDelayInSeconds > 0 &&
      directory !== EmailDirectory.Inbox
    ) {
      updateEmail({directory, receptionDelayInSeconds: 0}, () => {
        showMoveWarning()
        successFn()
      })
    } else {
      updateEmail({directory}, successFn)
    }
  }
  const markerOptions = React.useMemo<SelectOptionCustomized[]>(
    () => [
      {
        value: EmailStatus.Read,
        label: t("email__status_read")
      },
      {
        value: EmailStatus.Unread,
        label: t("email__status_unread"),
        renderOptionLabel: label => <EmailUnreadIndicator text={label} />
      }
    ],
    [email]
  )

  const showDelayOverlay = () => setDelayOverlayVisible(true)
  const hideDelayOverlay = () => setDelayOverlayVisible(false)

  return {
    directoryOptions: getDirectoryOptions(t),
    hideDelayOverlay,
    isDelayOverlayVisible,
    markerOptions,
    moveEmail,
    showDelayOverlay
  }
}
