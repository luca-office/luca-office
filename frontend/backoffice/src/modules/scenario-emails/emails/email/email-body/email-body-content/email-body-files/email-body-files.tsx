import * as React from "react"
import {Button, ColumnProps, DeleteOrArchiveEntityButton, Heading, Icon, UploadFileModal} from "shared/components"
import {HeadingLevel, IconName, UploadFileType} from "shared/enums"
import {Email} from "shared/models"
import {AddEmailFiles, EmailBodyFile} from "shared/office-tools/email-client/email-details/email-files"
import {CustomStyle, FontWeight, iconDefaultColor} from "shared/styles"
import {useLucaTranslation} from "shared/translations"
import {iconForFile} from "shared/utils"
import {Option} from "shared/utils/option"
import {emailStyle as styles} from "./email-body-files.style"
import {useEmailFiles} from "./hooks/use-email-files"

export interface EmailBodyFilesProps extends CustomStyle {
  readonly email: Email
  readonly disabled?: boolean
}

export const EmailBodyFiles: React.FC<EmailBodyFilesProps> = ({disabled = false, customStyles, email}) => {
  const {t} = useLucaTranslation()
  const {uploadVisible, showUpload, hideUpload, uploadBinaries, files, deleteFileHook, createTextDocumentFile} =
    useEmailFiles(email)

  const fileEntities: EmailBodyFile[] = files.map(file => ({
    id: file.id,
    title: file.name,
    relevance: Option.of(file.relevance),
    iconName: iconForFile(file)
  }))

  const columns: ColumnProps<EmailBodyFile>[] = [
    {
      key: "icon-title",
      customStyles: styles.iconTitleColumn,
      header: (
        <Heading fontWeight={FontWeight.Bold} level={HeadingLevel.h3}>
          {t("email__files_table_column_file")}
        </Heading>
      ),
      content: entity => (
        <div css={styles.nameColumnContent}>
          <Icon name={entity.iconName} />
          {entity.title}
        </div>
      )
    },
    {
      key: "create-file",
      customStyles: [styles.deleteColumn],
      header: (
        <Button
          disabled={disabled}
          icon={IconName.Add}
          title={t("email__files_table_column_file")}
          customStyles={styles.addButton}
          customIconStyles={styles.addButtonIcon}
          onClick={showUpload}
        />
      ),
      content: entity => (
        <DeleteOrArchiveEntityButton
          customStyles={styles.deleteButton}
          customButtonStyles={styles.deleteButtonInner}
          disabled={disabled}
          iconColor={iconDefaultColor}
          entityId={entity.id}
          useDeleteHook={deleteFileHook}
        />
      )
    }
  ]

  return (
    <>
      <AddEmailFiles files={fileEntities} columns={columns} customStyles={customStyles} />

      {uploadVisible && (
        <UploadFileModal
          acceptedFileTypes={[
            UploadFileType.Graphic,
            UploadFileType.Video,
            UploadFileType.PDF,
            UploadFileType.TableCalculation
          ]}
          onBinariesSuccessfullyUploaded={uploadBinaries}
          onDismiss={hideUpload}
          isLimitedToSingleItem={false}
          createTextDocument={createTextDocumentFile}
        />
      )}
    </>
  )
}
