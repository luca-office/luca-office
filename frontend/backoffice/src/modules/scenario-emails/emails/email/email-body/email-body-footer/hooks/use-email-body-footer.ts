import {useScenarioQuestionnaires} from "shared/graphql/hooks"
import {ScenarioQuestionnaire, SelectOption} from "shared/models"
import {useLucaTranslation} from "shared/translations"
import {getRelevanceCriteria} from "shared/utils"

export interface UseEmailBodyFooterHook {
  readonly relevanceOptions: SelectOption[]
  readonly scenarioQuestionnaires: ScenarioQuestionnaire[]
}

export const useEmailBodyFooter = (scenarioId: UUID): UseEmailBodyFooterHook => {
  const {t} = useLucaTranslation()

  const relevanceOptions = getRelevanceCriteria(t)

  const {scenarioQuestionnaires} = useScenarioQuestionnaires(scenarioId)

  return {
    relevanceOptions,
    scenarioQuestionnaires: scenarioQuestionnaires
      .getOrElse([])
      .filter(questionnaire => questionnaire.scenarioId === scenarioId)
  }
}
