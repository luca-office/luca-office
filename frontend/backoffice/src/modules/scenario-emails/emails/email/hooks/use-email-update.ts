import {EmailUpdate} from "shared/graphql/generated/globalTypes"
import {useUpdateEmail} from "shared/graphql/hooks"
import {Email} from "shared/models"
import {Option} from "shared/utils"

export interface UseEmailUpdateHook {
  readonly updateEmail: (update: Partial<EmailUpdate>, onSuccess?: () => void) => Promise<Option<Email>>
  readonly updateEmailLoading: boolean
}

export const useEmailUpdate = (email: Email): UseEmailUpdateHook => {
  const {updateEmail, updateEmailLoading} = useUpdateEmail()

  const emailUpdate = {
    sender: email.sender,
    recipient: email.recipient,
    ccRecipients: email.ccRecipients,
    subject: email.subject,
    directory: email.directory,
    receptionDelayInSeconds: email.receptionDelayInSeconds,
    isInitiallyRead: email.receptionDelayInSeconds >= 0 ? false : email.isInitiallyRead,
    relevance: email.relevance,
    message: email.message,
    scenarioId: email.scenarioId
  }

  const handleEmailUpdate = (update: Partial<EmailUpdate>, onSuccess?: () => void) =>
    updateEmail(email.id, {...emailUpdate, ...update}, onSuccess)

  return {
    updateEmail: handleEmailUpdate,
    updateEmailLoading
  }
}
