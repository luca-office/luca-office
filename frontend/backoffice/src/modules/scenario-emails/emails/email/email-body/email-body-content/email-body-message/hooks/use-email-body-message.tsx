import * as React from "react"
import {EmailMessageTemplate} from "shared/enums"
import {EmailUpdate} from "shared/graphql/generated/globalTypes"
import {useSampleCompanyLazy} from "shared/graphql/hooks"
import {Email, SampleCompany, Scenario} from "shared/models"
import {useLucaTranslation} from "shared/translations"
import {
  getCaretPosition,
  getEmailMessageTemplateValue,
  getEmailMessageWithTemplateLabels,
  Option,
  setCaretPosition
} from "shared/utils"
import {useEmailUpdate} from "../../../../hooks/use-email-update"

export interface useEmailBodyMessageHook {
  readonly dataLoading: boolean
  readonly sampleCompanyId: UUID | null
  readonly sampleCompany: Option<SampleCompany>
  readonly sampleCompanyName: string | null
  readonly sampleCompanyEmailSignature: string | null
  readonly emailMessage: string
  readonly updateEmailMessage: (value: string) => void
  readonly updateEmail: (update: EmailUpdate) => void
  readonly handleSubjectUpdate: (subject: string) => void
  readonly emailMessageHasSignature: boolean
  readonly emailMessageWithTemplateLabels: string
  readonly updateEmailMessageWithTemplate: (template: EmailMessageTemplate) => void
  readonly textAreaRef: React.MutableRefObject<HTMLTextAreaElement | null>
}

export const useEmailBodyMessage = (email: Email, scenario: Option<Scenario>): useEmailBodyMessageHook => {
  const {t} = useLucaTranslation()

  const textAreaRef = React.useRef<HTMLTextAreaElement | null>(null)

  const [emailMessage, updateEmailMessage] = React.useState(email.message)

  const sampleCompanyId = scenario.map(scenario => scenario.sampleCompanyId).orNull()

  const {sampleCompany, getSampleCompany} = useSampleCompanyLazy()
  const {updateEmail, updateEmailLoading} = useEmailUpdate(email)

  const emailMessageWithTemplateLabels = getEmailMessageWithTemplateLabels(t, email.message)

  const sampleCompanyName = sampleCompany.map(company => company.name).orNull()
  const sampleCompanyEmailSignature = sampleCompany.map(company => company.emailSignature).orNull()
  const emailMessageHasSignature =
    !!sampleCompanyEmailSignature && emailMessage.indexOf(sampleCompanyEmailSignature) !== -1

  const handleSubjectUpdate = (subject: string) =>
    new Promise<void>(resolve => updateEmail({subject: subject.trim()}, resolve))

  const updateEmailMessageWithTemplate = (template: EmailMessageTemplate) => {
    const templateValue = getEmailMessageTemplateValue(template)

    if (textAreaRef.current === null) {
      updateEmailMessage(emailMessage === "" ? templateValue : `${emailMessage} ${templateValue}`)
      return
    }

    const {start, end} = getCaretPosition(textAreaRef.current)

    updateEmailMessage(
      `${emailMessage.substring(0, start)}${templateValue}${emailMessage.substring(end, emailMessage.length)}`
    )

    const caretPosition = end + templateValue.length

    // use set timeout 0 to make sure the new value was already rendered
    setTimeout(
      () =>
        textAreaRef.current !== null &&
        setCaretPosition(textAreaRef.current, {start: caretPosition, end: caretPosition}),
      0
    )
  }

  React.useEffect(() => {
    if (sampleCompanyId !== null) {
      getSampleCompany(sampleCompanyId)
    }
  }, [sampleCompanyId])

  React.useEffect(() => {
    updateEmailMessage(email.message)
  }, [email.message])

  return {
    dataLoading: updateEmailLoading,
    sampleCompanyId,
    sampleCompany,
    sampleCompanyName,
    sampleCompanyEmailSignature,
    emailMessage,
    updateEmailMessage,
    updateEmail,
    handleSubjectUpdate,
    emailMessageHasSignature,
    emailMessageWithTemplateLabels,
    updateEmailMessageWithTemplate,
    textAreaRef
  }
}
