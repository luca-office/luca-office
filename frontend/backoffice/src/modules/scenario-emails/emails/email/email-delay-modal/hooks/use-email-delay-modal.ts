import {min} from "lodash"
import omit from "lodash-es/omit"
import * as React from "react"
import {useForm, UseFormMethods} from "react-hook-form"
import {TimeUnit} from "shared/enums"
import {useScenario} from "shared/graphql/hooks"
import {Email, EmailOpeningIntervention, SelectOption} from "shared/models"
import {useLucaTranslation} from "shared/translations"
import {Option, secondsToGivenTimeUnit} from "shared/utils"
import {MAX_RECEPTION_DELAY, receptionDelayOptionsConfig} from "../../../../../scenarios/config/delays-config"
import {useEmailUpdate} from "../../hooks/use-email-update"
import {getReceptionDelayInSeconds} from "../email-delay-modal-utils"

export enum EmailTemporalStage {
  Past = "Past",
  Present = "Present",
  Future = "Future"
}

export type EmailDelayTimeUnit = TimeUnit.Hour | TimeUnit.Minute | TimeUnit.Second
export interface EmailDelayForm {
  readonly receptionDelay: number
  readonly timeUnit: EmailDelayTimeUnit
}

export interface UseEmailDelayModalHook {
  readonly formMethods: UseFormMethods<EmailDelayForm>
  readonly maxFutureDelay: Option<number>
  readonly receptionDelayOptions: SelectOption[]
  readonly updateEmailDelay: (data: EmailDelayForm) => void
  readonly selectedTemporalStage: EmailTemporalStage
  readonly toggleTemporalStage: (stage: EmailTemporalStage) => void
}

export const useEmailDelayModal = (
  email: Email,
  emailOpeningInterventions: EmailOpeningIntervention[],
  dismiss?: () => void
): UseEmailDelayModalHook => {
  const isSendInFuture = email.receptionDelayInSeconds > 0

  const isSendImmediately = email.receptionDelayInSeconds === 0

  const defaultSelectedTemporalStage = isSendInFuture
    ? EmailTemporalStage.Future
    : isSendImmediately
    ? EmailTemporalStage.Present
    : EmailTemporalStage.Past

  const {t} = useLucaTranslation()

  const [selectedTemporalStage, setSelectedTemporalStages] = React.useState<EmailTemporalStage>(
    defaultSelectedTemporalStage
  )

  const {updateEmail} = useEmailUpdate(email)
  const {scenario} = useScenario(email.scenarioId)

  const defaultTimeUnit = getDefaultTimeUnit(email.receptionDelayInSeconds)
  const defaultReceptionDelay = Math.abs(secondsToGivenTimeUnit(defaultTimeUnit, email.receptionDelayInSeconds))
  const isFutureTemporalStageSelected = selectedTemporalStage === EmailTemporalStage.Future

  const formMethods = useForm<EmailDelayForm>({
    defaultValues: {
      receptionDelay: defaultReceptionDelay,
      timeUnit: defaultTimeUnit
    }
  })

  const maxFutureDelay = React.useMemo<Option<number>>(
    () => scenario.map(scenarioData => scenarioData.maxDurationInSeconds || 0),
    [scenario]
  )

  const minimumDelayOfEmailOpeningInterventions = min(
    emailOpeningInterventions.map(intervention => intervention.timeOffsetInSeconds)
  )

  const receptionDelayOptions = React.useMemo<SelectOption[]>(
    () => receptionDelayOptionsConfig.map(config => ({...omit(config, "labelKey"), label: t(config.labelKey)})),
    []
  )

  const toggleTemporalStage = (stage: EmailTemporalStage) => {
    if (stage === EmailTemporalStage.Present) {
      formMethods.setValue("receptionDelay", 0)
    }
    setSelectedTemporalStages(stage)
  }

  const updateEmailDelay = (form: EmailDelayForm) => {
    const receptionDelay = isFutureTemporalStageSelected
      ? getReceptionDelayInSeconds(form.timeUnit, form.receptionDelay)
      : -getReceptionDelayInSeconds(form.timeUnit, form.receptionDelay)

    if (Math.abs(receptionDelay) > MAX_RECEPTION_DELAY) {
      formMethods.setError("receptionDelay", {
        shouldFocus: true,
        message: t("validation__email_delay_max_error")
      })
    } else if (
      isFutureTemporalStageSelected &&
      minimumDelayOfEmailOpeningInterventions !== undefined &&
      receptionDelay > minimumDelayOfEmailOpeningInterventions
    ) {
      formMethods.setError("receptionDelay", {
        shouldFocus: true,
        message: t("validation__email_delay_intervention_email")
      })
    } else if (isFutureTemporalStageSelected && receptionDelay > maxFutureDelay.getOrElse(MAX_RECEPTION_DELAY)) {
      formMethods.setError("receptionDelay", {
        shouldFocus: true,
        message: t("validation__email_delay")
      })
    } else {
      updateEmail(
        {
          receptionDelayInSeconds: selectedTemporalStage === EmailTemporalStage.Present ? 0 : receptionDelay,
          ...(Math.sign(receptionDelay) === 1 ? {isInitiallyRead: false} : undefined)
        },
        dismiss
      )
    }
  }

  return {
    formMethods,
    maxFutureDelay,
    receptionDelayOptions,
    selectedTemporalStage,
    toggleTemporalStage,
    updateEmailDelay
  }
}

export const getDefaultTimeUnit = (receptionDelay: number): EmailDelayTimeUnit => {
  if (receptionDelay === 0) {
    return TimeUnit.Second
  } else {
    return receptionDelay % 3600 === 0 ? TimeUnit.Hour : receptionDelay % 60 === 0 ? TimeUnit.Minute : TimeUnit.Second
  }
}
