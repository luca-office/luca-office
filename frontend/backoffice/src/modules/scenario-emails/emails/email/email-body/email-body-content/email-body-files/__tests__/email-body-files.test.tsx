import {fireEvent, render, screen} from "@testing-library/react"
import {mount} from "enzyme"
import * as React from "react"
import {act, create} from "react-test-renderer"
import {Button, DeleteOrArchiveEntityButton, TableContainer, Tooltip} from "shared/components"
import {filesMock} from "shared/graphql/__mocks__"
import {deleteEntityHookMock} from "../../../../../../../../graphql/__mocks__"
import {emailsMock} from "../../../../../hooks/__mocks__/emails.mock"
import {EmailBodyFiles, EmailBodyFilesProps} from "../email-body-files"
import * as useEmailFilesHook from "../hooks/use-email-files"
import {UseEmailFilesHook} from "../hooks/use-email-files"

const email = emailsMock[0]

const defaultProps: EmailBodyFilesProps = {
  disabled: false,
  email
}

const emailBodyFilesHookValuesDefault: UseEmailFilesHook = {
  uploadVisible: false,
  files: [],
  hideUpload: jest.fn(),
  showUpload: jest.fn(),
  filesLoading: false,
  uploadBinaries: jest.fn(),
  createTextDocumentFile: jest.fn(),
  deleteFileHook: () => deleteEntityHookMock
}

const emailFilesSpy = jest.spyOn(useEmailFilesHook, "useEmailFiles")

const getComponent = (props?: Partial<EmailBodyFilesProps>) => <EmailBodyFiles {...{...defaultProps, ...props}} />

describe("email-body-files", () => {
  it("renders correctly", () => {
    emailFilesSpy.mockReturnValue(emailBodyFilesHookValuesDefault)
    const component = getComponent()
    const tree = create(component).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it("adds binaries", () => {
    const showModalMock = jest.fn()
    emailFilesSpy.mockReturnValue({
      ...emailBodyFilesHookValuesDefault,
      showUpload: showModalMock,
      files: filesMock
    })

    const component = getComponent()
    const tree = mount(component)

    render(component)

    const table = tree.find(TableContainer)
    const addBtn = table.find(Button).first()

    expect(addBtn.prop("disabled")).toBe(false)
    act(() => {
      fireEvent.click(screen.getByTitle("email__files_table_column_file"))
    })

    expect(showModalMock).toHaveBeenCalledTimes(1)
    expect(table).toHaveLength(1)
    const rows = table.find(".entity-row")
  })
})
