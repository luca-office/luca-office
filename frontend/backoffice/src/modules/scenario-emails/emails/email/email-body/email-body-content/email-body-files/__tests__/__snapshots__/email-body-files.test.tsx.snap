// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`email-body-files renders correctly 1`] = `
<div
  className="css-1eq9cru-AddEmailFiles"
>
  <div
    className="css-3vqmyq-AddEmailFiles"
  >
    <div
      className="label css-ksst2g-AddEmailFiles"
    >
      <div
        className="css-ln7jn1-Heading"
      >
        email__files_label (0)
      </div>
      <div
        className="css-0"
        id="tooltip_42"
      >
        <div
          className="css-l0jg45-Icon"
          dangerouslySetInnerHTML={
            {
              "__html": {},
            }
          }
        />
      </div>
      <div
        className="react-tooltip styles-module_tooltip__mnnfp styles-module_dark__xNqje css-10zyuev-Tooltip"
        role="tooltip"
        style={
          {
            "backgroundColor": "#2e3032",
            "zIndex": 1,
          }
        }
      >
        <div
          className="css-1tdk34p-Tooltip"
        >
          <div
            className="css-1kjzr28-Tooltip"
          >
            <div
              className="css-jb7kz7-Heading"
            >
              email__files_label
            </div>
            <div
              className="css-8tih5m-Text"
            >
              email__files_tooltip
            </div>
          </div>
          <div
            className="css-1fvvs0r-Tooltip"
          >
            <div
              className="css-nlq2eo-Icon"
              dangerouslySetInnerHTML={
                {
                  "__html": {},
                }
              }
            />
          </div>
        </div>
        <div
          className="react-tooltip-arrow styles-module_arrow__K0L3T"
          style={{}}
        />
      </div>
    </div>
    <div
      className="css-1r45p5k-Table"
    >
      <div
        className="css-1r9mhfr-Table"
      >
        <div
          className="css-obhy4-Table"
        >
          <div
            className="css-8050ac-Heading"
          >
            email__files_table_column_file
          </div>
        </div>
        <div
          className="css-5ye7b5-Table"
        >
          <button
            className="css-23po1d-Button"
            disabled={false}
            onClick={[Function]}
            title="email__files_table_column_file"
          >
            <div
              className="css-95j2hm-Button"
            >
              <div
                className="css-xjsa35-Icon"
                dangerouslySetInnerHTML={
                  {
                    "__html": {},
                  }
                }
              />
            </div>
          </button>
        </div>
      </div>
      <div
        className="css-oyu5wv-Table"
      >
        <div
          className="table-placeholder css-mp88b7-Table"
        >
          email__files_placeholder
        </div>
      </div>
    </div>
  </div>
</div>
`;
